tensorboard/data/server/server.rs

```diff
@@ -92,44 +93,122 @@ impl TensorBoardDataProvider for DataProviderHandler {
 
     async fn list_scalars(
         &self,
-        _request: Request<data::ListScalarsRequest>,
+        req: Request<data::ListScalarsRequest>,
     ) -> Result<Response<data::ListScalarsResponse>, Status> {
+        let req = req.into_inner();
+        let want_plugin = parse_plugin_filter(req.plugin_filter)?;
+        let (run_filter, tag_filter) = parse_rtf(req.run_tag_filter);
+        let runs = self.read_runs()?;
+
         let mut res: data::ListScalarsResponse = Default::default();
-        let mut run: data::list_scalars_response::RunEntry = Default::default();
-        run.run_name = "train".to_string();
-        run.tags.push(data::list_scalars_response::TagEntry {
-            tag_name: "accuracy".to_string(),
-            metadata: Some(data::ScalarMetadata {
-                max_step: 5,
-                max_wall_time: FAKE_START_TIME + 5.0,
-                ..Default::default()
-            }),
-            ..Default::default()
-        });
-        res.runs.push(run);
+        for (run, data) in runs.iter() {
+            if !run_filter.want(run) {
```

Comment:
FWIW, iterating over the filter instead of the keyset would seem likely to save effort on net (assuming single-run lookups against a big experiment are fairly plausible, while sending a filter with many non-existent runs is unlikely; I guess most optimal would be to iterate over the shorter one).  Thoughts?  It probably does make the code a little messier so maybe better to wait.

Reply:
Okay, on thinking a bit more I guess a simpler formulation is


```rust
impl<T: Hash + Eq> Filter<T> {
    fn pluck<'f, 'm: 'f, V: 'm>(
        &'f self,
        it: impl Iterator<Item = (&'m T, &'m V)> + 'm,
    ) -> impl Iterator<Item = (&'m T, &'m V)> + 'f
    where
        T: Eq + Hash + 'm,
    {
        let which: Option<&'f HashSet<T>> = match self {
            Filter::All => None,
            Filter::Just(which) => Some(which),
        };
        it.filter(move |(k, _)| match which {
            None => true,
            Some(which) => which.contains(k),
        })
    }
}
```

which should about have the same runtime properties as what I sketched
above: the `match self` on `Either::next` in that version corresponds to
the `match which` in the `it.filter` closure in this one.

So less complicated (almost fits on a screen), but the lifetime
annotations are still a bit much. But I could at least reasonably be
convinced now. Let me know if you have any thoughts.

Reply:
Thanks for the detailed thoughts!  Didn't mean to make this too much of a digression, totally fine to leave as just a potential TODO for now.  Maybe if I have any particularly good ideas I'll take a stab at this myself :)

Reply:
Yep, TODO included. And, sure, feel free to grab it. :-)